docs/cram-tests/index.md

```diff
@@ -0,0 +1,368 @@
+`melange.emit` stanzas are automatically attached to the `@melange` alias.
+
+:::
+
+## Promotion
+
+After running `opam exec -- dune build @runtest`, you should be able to see the
+output from the cram test you just added. However, this doesn't count as a
+passing cram test, because this output can't be compared to the expected output.
+Dune has a nice feature that lets you
+[promote](https://dune.readthedocs.io/en/stable/concepts/promotion.html#promotion)
+the latest output of a cram test to become the expected output:
+
+```shell
+opam exec -- dune promote
+```
+
+The above command updates your `src/order-confirmation/tests.t` cram test file,
+which now looks something like this:
+
+<<< tests{cram}
+
+The part underneath the command is the expected output of the command.
+
+## Add `npm` scripts
+
+Instead of typing `opam exec` commands over and over, let's add the relevant
+`npm` scripts to your `package.json` file:
+
+```json
+"test": "npm run build -- @runtest",
+"testwatch": "npm run build -- @runtest --watch",
+"promote": "npm run dune -- promote"
+```
+
+You should also add some lines to the `scriptsComments` section of
+`package.json`:
+
+```json
+"test": "# Run the tests",
+"testwatch": "# Watch files and re-run tests",
+"promote": "# Promote most recent output to expected output"
+```
+
+Run `npm run test` to check that it works.
+
+## Sanitize cram test output
+
+The cram test still doesn't succeed, but now it's because the most recent output
+and the expected output don't match:
+
+<<< diff{diff}
+
+The problem is that the recorded durations for individual unit tests is differs
+on each test run. Therefore we must
+[sanitize](https://dune.readthedocs.io/en/stable/tests.html#test-output-sanitation)
+the test output, i.e. remove the parts that are non-deterministic. Update your
+cram test command:
+
+```shell
+$ node ./output/src/order-confirmation/SandwichTests.mjs | sed '/duration_ms/d'
+```
+
+Now the output of the command is piped to `sed '/duration_ms/d'`, which removes
+all the lines containing the string `duration_ms`.
+
+Follow these steps to achieve your first successful cram test run:
+
+1. Run `npm run test`
+1. It fails because the expected output still contains the `duration_ms` lines
+1. Run `npm run promote` to promote the latest output to the expected output
+1. Run `npm run test`
+1. You should see no test output
+
+No test output means it succeeded. Somewhat unintuitively, when cram tests
+succeed, they remain silent![^1]
+
+## Sandbox
+
+All cram tests execute in a
+[sandbox](https://dune.readthedocs.io/en/stable/tests.html#sandboxing). What
+goes into the sandbox directory is determined by the value of the `cram/deps`
+field in your `dune` file. Let's write an explorative cram test that allows us
+to peek inside the sandbox. First, install the [tree-node-cli
+package](https://www.npmjs.com/package/tree-node-cli):
+
+```shell
+npm install --save-dev tree-node-cli
```

Comment:
I went with `tree-node-cli` because MacOS doesn't have the `tree` command, and even if you install one via `brew install tree`, it doesn't seem to be the same command as on Linux. Even on Linux, which does have the `tree` command, the output doesn't look very nice when run by a cram test.